### ISSUE TITLE
train/gen_kl_error persistently high (0.15–1.75) during GRPO training with NeMo Gym VLM support

### ISSUE THREAD

**Issue by shashank3959** [NVIDIA-NeMo/RL#2260]
**Describe the bug**

`gen_kl_error` stays between 0.15–0.65 (with spikes to 1.75) throughout a full GRPO training run on `circle_click` with `Qwen/Qwen3-VL-2B-Instruct`. This metric measures `KL(P_gen || P_train)` — the divergence between logprobs from vLLM inference and logprobs from the DTensor training forward pass on the **same model weights and same token sequences**.

Training still converges (95% val accuracy), but the noisy importance-sampling ratio may hurt sample efficiency on harder tasks.
                                                                                                                                                                                                                                                                 
**Observed values across 62 training steps:**

| Phase | Typical `gen_kl_error` | Range |
|-------|----------------------|-------|
| Steps 1–20 (early) | ~0.45 | 0.32 – 0.64 |
| Steps 21–40 (mid) | ~0.30 | 0.16 – 1.75 (spike at step 22) |
| Steps 41–62 (late) | ~0.20 | 0.14 – 0.60 |

<img width="1036" height="751" alt="Image" src="https://github.com/user-attachments/assets/3a292ff8-bcd4-41ce-905a-52961fe27c55" />

Despite the high KL error, reward and validation accuracy still converge (val accuracy reaches 0.95, avg reward ~1.0), so training is not broken — but the elevated gen_kl_error suggests the GRPO importance-sampling ratio (`exp(curr_logprobs - generation_logprobs)`) used in the clipped PG loss is computed from a noisy baseline, which may reduce sample efficiency or cause instability in harder tasks.

## Steps/Code to reproduce bug

```bash
# 1. Clone and checkout PR #2092
git clone git@github.com:NVIDIA-NeMo/RL.git nemo-rl --recursive
cd nemo-rl
git fetch origin pull/2092/head:cmunley1/gym-vlm
git checkout cmunley1/gym-vlm
git submodule update --init --recursive

# 2. Set up venvs and generate data
uv venv
cd 3rdparty/Gym-workspace/Gym
uv venv --python 3.12 --allow-existing .venv
source .venv/bin/activate
SETUPTOOLS_SCM_PRETEND_VERSION=0.0.0 uv sync --active --extra dev
mkdir -p data/circle_click
python3 resources_servers/circle_click/generate_data.py --n 1000 --out data/circle_click/train.jsonl
python3 resources_servers/circle_click/generate_data.py --n 100  --out data/circle_click/validation.jsonl
deactivate
cd ../../..

# 3. Run GRPO training (4x H100, single node)
export CUDNN_HOME=.venv/lib/python3.12/site-packages/nvidia/cudnn
export LD_LIBRARY_PATH=".venv/lib/python3.12/site-packages/nvidia/cudnn/lib:${LD_LIBRARY_PATH:-}"
export TORCH_CUDA_ARCH_LIST="9.0"

uv run python examples/nemo_gym/run_grpo_nemo_gym.py \
  --config examples/nemo_gym/grpo_circle_click_qwen3vl2b.yaml \
  cluster.num_nodes=1 \
  cluster.gpus_per_node=4
```

Uses the default config `grpo_circle_click_qwen3vl2b.yaml` shipped with the PR — no overrides beyond node/GPU count.


**Expected behavior**

Expected ideal values of **gen_kl_error** are near zero (< 0.01). The consistently elevated values suggest a systematic numerical divergence between the two inference backends.

**Comment by vinhngx** [comment]
I've observed a few intriguing behaviours of the Qwen/Qwen3-VL-2B-Instruct model.

So this
```
python3 resources_servers/circle_click/generate_data.py --n 1000 --out data/circle_click/train.jsonl
python3 resources_servers/circle_click/generate_data.py --n 100  --out data/circle_click/validation.jsonl
```
will generate a screen size of 1000x1000 by default. If we just validate the accuracy with nemo gym roll out collect, the base model accuracy is already ~98%.

However, if we expand the screen size to include varying sizes e.g. `--img-size-min 384 --img-size-max 1600`, the accuracy in the non-1k-by-1k buckets is low, like in the 30%. What's more intriguing is that the click co-ordinates produced by the model seem to be anchored to the canonical 1k x 1k size. If we rescale those co-ordinates to the actual screen size, the accuracy would be near perfect again.
So essentially, with RL, we want to de-anchor the model from that behaviour. I'm curious to see what the accuracy on varying-sized datasets is like (from what I tried, accuracy wouldn't budge after prolonged training).

**Comment by vinhngx** [comment]
Some more interesting observations on this task when evaluating the Qwen3-VL-2B-instruct model with vLLM with and without tool call enforcement

## TL;DR
The base Qwen/Qwen3-VL-2B-Instruct model appears to emit click coordinates in an implicit 1000x1000 coordinate frame, even when the prompt states a different image size. Directly scoring those coordinates gives poor accuracy off the 1000x1000 bucket, but rescaling from 1000x1000 to the actual image size substantially improves results.

Forcing tool use makes the behavior much clearer: with tool_choice="required", rescaled accuracy rises to 95% overall, versus 41.88% with auto tool choice. This suggests the model can localize the target well, but reliable tool emission and coordinate-frame normalization are both important for evaluation.


## Summary

We evaluated the base `Qwen/Qwen3-VL-2B-Instruct` model on a synthetic circle-click task with square screen sizes from `300x300` to `1500x1500`, including `1000x1000`. Each bucket contains 20 examples.

Two decoding modes were compared:

- Auto tool choice: `tool_choice = "auto"`
- Forced tool choice: `tool_choice = "required"` with a single available `click` tool

Both runs used vLLM with `--enable-auto-tool-choice --tool-call-parser hermes`. The model was evaluated from the base `Qwen/Qwen3-VL-2B-Instruct` weights.

## Evaluation Setup

The prompt explicitly names the image size, for example:

```text
This is a 1000x1000 pixel image. It shows ..., and a <color> circle. Click on the <color> circle.
```

Each response is verified two ways:

- Direct coordinates: use the model's `x, y` as coordinates in the actual image.
- Rescaled-from-1000 coordinates: interpret the model's `x, y` as coordinates in a `1000x1000` frame, then map to the actual image with:

```text
rescaled_x = clicked_x * actual_size / 1000
rescaled_y = clicked_y * actual_size / 1000
```

## Aggregate Results

| mode | direct hits | direct accuracy | rescaled hits | rescaled accuracy |
|---|---:|---:|---:|---:|
| Auto tool choice | 21/160 | 13.13% | 67/160 | 41.88% |
| Forced tool choice | 45/160 | 28.12% | 152/160 | 95.00% |

## Per-Bucket Results

| screen size | auto direct | auto rescaled | forced direct | forced rescaled |
|---:|---:|---:|---:|---:|
| 300 | 0/20 | 5/20 | 0/20 | 15/20 |
| 500 | 0/20 | 10/20 | 0/20 | 19/20 |
| 700 | 0/20 | 10/20 | 0/20 | 20/20 |
| 900 | 3/20 | 9/20 | 7/20 | 19/20 |
| 1000 | 10/20 | 10/20 | 20/20 | 20/20 |
| 1100 | 8/20 | 11/20 | 15/20 | 19/20 |
| 1300 | 0/20 | 6/20 | 1/20 | 20/20 |
| 1500 | 0/20 | 6/20 | 2/20 | 20/20 |

## Interpretation

The model appears to prefer emitting click coordinates in an implicit `1000x1000` coordinate frame. Direct-coordinate evaluation performs poorly away from the `1000x1000` bucket, while rescaling predictions from a `1000x1000` frame substantially improves accuracy.

Forcing tool choice is also important. With `tool_choice = "required"`, the rescaled-from-1000 accuracy reaches `95.00%`, compared with `41.88%` for auto tool choice. At `1000x1000`, forced tool choice gives `20/20` direct accuracy, which supports the hypothesis that the model can solve the visual localization task when tool emission is reliable and the coordinate frame matches its implicit convention.
